software_scale_context.go

```diff
@@ -0,0 +1,171 @@
+package astiav
+
+//#cgo pkg-config: libswscale
+//#include <libswscale/swscale.h>
+import "C"
+import (
+	"fmt"
+)
+
+// https://github.com/FFmpeg/FFmpeg/blob/n5.0/libswscale/swscale_internal.h#L300
+type SoftwareScaleContext struct {
+	c *C.struct_SwsContext
+	// We need to store attributes in GO since C attributes are internal and therefore not accessible
+	dstFormat C.enum_AVPixelFormat
+	dstH      C.int
+	dstW      C.int
+	flags     SoftwareScaleContextFlags
+	srcFormat C.enum_AVPixelFormat
+	srcH      C.int
+	srcW      C.int
+}
+
+func NewSoftwareScaleContext(srcW, srcH int, srcFormat PixelFormat, dstW, dstH int, dstFormat PixelFormat, flags SoftwareScaleContextFlags) *SoftwareScaleContext {
+	ssc := SoftwareScaleContext{
+		dstFormat: C.enum_AVPixelFormat(dstFormat),
+		dstH:      C.int(dstH),
+		dstW:      C.int(dstW),
+		flags:     flags,
+		srcFormat: C.enum_AVPixelFormat(srcFormat),
+		srcH:      C.int(srcH),
+		srcW:      C.int(srcW),
+	}
+
+	ssc.c = C.sws_getContext(
+		ssc.srcW,
+		ssc.srcH,
+		ssc.srcFormat,
+		ssc.dstW,
+		ssc.dstH,
+		ssc.dstFormat,
+		C.int(ssc.flags),
+		nil, nil, nil,
+	)
+	if ssc.c == nil {
+		return nil
+	}
+	return &ssc
+}
+
+func (ssc *SoftwareScaleContext) ScaleFrame(src, dst *Frame) (height int) {
```

Comment:
OK could just short explain it why this is better ?

Reply:
Oh its Just any easier interface

Reply:
> Oh its Just any easier interface

Exactly, and with it we don't need to set height, width and pixel format, and allocate buffer of the destination frame since it's done [automatically](https://ffmpeg.org/doxygen/5.1/swscale_8c_source.html#l01087) which should take care of one of your previous comment.

Also it seems to handle multithreading when `sws_scale` doesn't 🤔

Reply:
great if its easier and more practical